modules/shared_vpc_access/main.tf

```diff
@@ -19,40 +19,49 @@ data "google_project" "service_project" {
 }
 
 locals {
-  gke_shared_vpc_enabled = contains(var.active_apis, "container.googleapis.com")
-  gke_s_account = local.gke_shared_vpc_enabled ? format(
-    "service-%s@container-engine-robot.iam.gserviceaccount.com",
-    data.google_project.service_project.number,
-  ) : ""
+  apis = {
+    "container.googleapis.com" : format("service-%s@container-engine-robot.iam.gserviceaccount.com", data.google_project.service_project.number),
+    "dataproc.googleapis.com" : format("service-%s@dataproc-accounts.iam.gserviceaccount.com", data.google_project.service_project.number),
+  }
+  gke_shared_vpc_enabled      = contains(var.active_apis, "container.googleapis.com")
   dataproc_shared_vpc_enabled = contains(var.active_apis, "dataproc.googleapis.com")
-  dataproc_s_account = local.dataproc_shared_vpc_enabled ? format(
-    "service-%s@dataproc-accounts.iam.gserviceaccount.com",
-    data.google_project.service_project.number
-  ) : ""
-  active_api_s_accounts = compact([local.gke_s_account, local.dataproc_s_account])
+  active_apis                 = setintersection(keys(local.apis), var.active_apis)
+  subnetwork_api              = length(var.shared_vpc_subnets) != 0 ? tolist(setproduct(local.active_apis, var.shared_vpc_subnets)) : []
 }
 
 /******************************************
-  compute.networkUser role granted to GKE service account for GKE on shared VPC subnets
+  if "container.googleapis.com" compute.networkUser role granted to GKE service account for GKE on shared VPC subnets
+  if "dataproc.googleapis.com" compute.networkUser role granted to dataproc service account for dataproc on shared VPC subnets
   See: https://cloud.google.com/kubernetes-engine/docs/how-to/cluster-shared-vpc
  *****************************************/
-resource "google_compute_subnetwork_iam_member" "gke_shared_vpc_subnets" {
+resource "google_compute_subnetwork_iam_member" "gke_dataproc_shared_vpc_subnets" {
   provider = google-beta
-  count    = local.gke_shared_vpc_enabled && length(var.shared_vpc_subnets) != 0 ? length(var.shared_vpc_subnets) : 0
+  count    = length(var.shared_vpc_subnets) != 0 && (local.gke_shared_vpc_enabled || local.dataproc_shared_vpc_enabled) ? length(local.subnetwork_api) : 0
```

Comment:
Can we change this to `for_each = local.subnetwork_api`?

Reply:
yeah I tried this yesteday, but ran into the `cannot determine number of resources error`. I will give it another whirl with new structure.

Reply:
So I think the problem with for_each is that at plan time we only know the length and not the values. Similar to this issue here https://github.com/hashicorp/terraform/issues/22735#issuecomment-580491563
Let me know if you have other suggestions.

I tried something like
```hcl
locals{
subnetwork_api = [for pair in setproduct(var.active_apis, var.shared_vpc_subnets): {
    api = pair[0]
    subnet = pair[1]
    }
  ] : []
}
 resource "google_compute_subnetwork_iam_member" "gke_dataproc_shared_vpc_subnets" {
...
for_each    = {for pair in local.subnetwork_api: "${pair.api}.${pair.subnet}" => pair}
}
```

Reply:
But we should know the values of the APIs though, no?

Reply:
Yeah I think its the subnets_self_links since they are created dynamically
https://github.com/terraform-google-modules/terraform-google-project-factory/blob/9eb64e2217bef8477ae07e1a834c4bfb3f64273f/examples/shared_vpc/main.tf#L118

Reply:
Hmm ok I guess we can stick to count then.